assets/php/class-wgpb-block-grid-base.php

```diff
@@ -240,64 +238,118 @@ protected function get_container_classes() {
 	 * @return string Rendered product output.
 	 */
 	public function render_product( $id ) {
-		global $product;
 		$product = wc_get_product( $id );
+
 		if ( ! $product ) {
 			return '';
 		}
 
-		$title = sprintf(
-			'<div class="wc-block-grid__product-title">%s</div>',
-			wp_kses_post( $product->get_title() )
-		);
-
-		$rating_str = $this->get_rating( $product );
-
-		$price_str = sprintf(
-			'<div class="wc-block-grid__product-price price">%s</div>',
-			$product->get_price_html()
+		$data = (object) array(
+			'permalink' => $product->get_permalink(),
+			'image'     => $this->get_image_html( $product ),
+			'title'     => $this->get_title_html( $product ),
+			'rating'    => $this->get_rating_html( $product ),
+			'price'     => $this->get_price_html( $product ),
+			'button'    => $this->get_button_html( $product ),
 		);
 
-		ob_start();
-		woocommerce_show_product_sale_flash();
-		$on_sale_flag = ob_get_contents();
-		ob_end_clean();
-
-		$button_str = sprintf(
-			'<div class="wc-block-grid__product-add-to-cart">%s</div>',
-			$this->get_add_to_cart( $product )
-		);
+		return "
+			<li class=\"wc-block-grid__product product\">
+				<a href=\"{$data->permalink}\" class=\"wc-block-grid__product-link\">
+					{$data->image}
+					{$data->title}
+					{$data->rating}
+					{$data->price}
+				</a>
+				{$data->button}
+			</li>
+		";
+	}
 
-		$content = sprintf(
-			'<a href="%s" class="wc-block-grid__product-link">',
-			$product->get_permalink()
-		);
+	/**
+	 * Get the product image.
+	 *
+	 * @param WC_Product $product Product.
+	 * @return string
+	 */
+	protected function get_image_html( $product ) {
+		return '<div class="wc-block-grid__product-image">' . $product->get_image( 'woocommerce_thumbnail' ) . '</div>';
+	}
 
-		$content .= sprintf(
-			'<div class="wc-block-grid__product-image">%s</div>',
-			woocommerce_get_product_thumbnail()
-		);
+	/**
+	 * Get the product title.
+	 *
+	 * @param WC_Product $product Product.
+	 * @return string
+	 */
+	protected function get_title_html( $product ) {
+		if ( empty( $this->attributes['contentVisibility']['title'] ) ) {
```

Comment:
Why `empty` and not just a simple boolean check? (just curious, this works fine)

Reply:
`empty` prevents a notice if `contentVisibility` or `title` is missing from the settings array. I'm not sure how likely that is to happen though.